library/core/src/future/async_drop.rs

```diff
@@ -0,0 +1,271 @@
+}
+
+/// Creates the asynchronous destructor of the pointed-to value.
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `to_drop` must be [valid](crate::ptr#safety) for both reads and writes.
+///
+/// * `to_drop` must be properly aligned, even if `T` has size 0.
+///
+/// * `to_drop` must be nonnull, even if `T` has size 0.
+///
+/// * The value `to_drop` points to must be valid for async dropping,
+///   which may mean it must uphold additional invariants. These
+///   invariants depend on the type of the value being dropped. For
+///   instance, when dropping a Box, the box's pointer to the heap must
+///   be valid.
+///
+/// * While `async_drop_in_place` is executing or the returned async
+///   destructor is alive, the only way to access parts of `to_drop`
+///   is through the `self: Pin<&mut Self>` references supplied to
+///   the `AsyncDrop::async_drop` methods that `async_drop_in_place`
+///   or `AsyncDropInPlace<T>::poll` invokes. This usually means the
+///   returned future stores the `to_drop` pointer and user is required
+///   to guarantee that dropped value doesn't move.
+///
+#[unstable(feature = "async_drop", issue = "none")]
+pub unsafe fn async_drop_in_place<T: ?Sized>(to_drop: *mut T) -> AsyncDropInPlace<T> {
+    // SAFETY: `async_drop_in_place_raw` has the same safety requirements
+    unsafe { AsyncDropInPlace(async_drop_in_place_raw(to_drop)) }
+}
+
+/// A future returned by the [`async_drop_in_place`].
+#[unstable(feature = "async_drop", issue = "none")]
+pub struct AsyncDropInPlace<T: ?Sized>(<T as AsyncDestruct>::AsyncDestructor);
+
+#[unstable(feature = "async_drop", issue = "none")]
+impl<T: ?Sized> fmt::Debug for AsyncDropInPlace<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AsyncDropInPlace").finish_non_exhaustive()
+    }
+}
+
+#[unstable(feature = "async_drop", issue = "none")]
+impl<T: ?Sized> Future for AsyncDropInPlace<T> {
+    type Output = ();
+
+    #[inline(always)]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // SAFETY: This code simply forwards poll call to the inner future
+        unsafe { Pin::new_unchecked(&mut self.get_unchecked_mut().0) }.poll(cx)
+    }
+}
+
+// FIXME(zetanumbers): Add same restrictions on AsyncDrop impls as
+//   with Drop impls
+/// Custom code within the asynchronous destructor.
+#[unstable(feature = "async_drop", issue = "none")]
+#[lang = "async_drop"]
+pub trait AsyncDrop {
+    /// A future returned by the [`AsyncDrop::async_drop`] to be part
+    /// of the async destructor.
+    #[unstable(feature = "async_drop", issue = "none")]
+    type Dropper<'a>: Future<Output = ()>
+    where
+        Self: 'a;
+
+    /// Constructs the asynchronous destructor for this type.
+    #[unstable(feature = "async_drop", issue = "none")]
+    fn async_drop(self: Pin<&mut Self>) -> Self::Dropper<'_>;
```

Comment:
This is something I should have wondered about much earlier, but I never saw any rationale (neither in the MCP, nor in the explainer for this PR), that weighs returning a future here against making this function a `poll_drop` function.

I think it may end up much less complex if we don't need to bounce back and forth between futures and things creating futures. `async_drop_in_place_raw` can then also not return a future, but just process the elements in place.

Unless there's a good reason for the future-returning variant that I am missing, I think it would be important to compare the two designs. Normally I would just go ahead and try out an alternative instead of coming with such a major redesign into a PR, but I am completely busy with impl-trait-in-associated-types at the moment. Do you have the capacity to try this out? Doesn't need to be the full thing you did in this PR, as long as a simple struct containing an `AsyncDrop` field works.

Reply:
> I think it may end up much less complex if we don't need to bounce back and forth between futures and things creating futures. `async_drop_in_place_raw` can then also not return a future, but just process the elements in place.

I believe this complexity may be warranted. To provide one example: `fn poll_drop` requires that the drop state is placed inside of the struct we implement it on, because it doesn't return a new future to hold the state. Types such as `Vec` in the stdlib have a stable layout and cannot be updated to contain more state. Dealing with `Vec<impl Future>` is not uncommon; and we should want it to be able to forward async destructors.

```rust
// `Vec` in the stdlib provides a stable layout
struct Vec<T> { .. }

// ❌ cannot or cannot efficiently implement `AsyncDrop::poll_drop`
// because we cannot track the drop state as part of `Vec`.
impl<T> AsyncDrop for Vec<T> {
    fn poll_drop(self: Pin<&mut Self>, cx: ...);
}

// ✅ we can track the drop state as part of the future
// returned by `AsyncDrop`, providing a location to track
// the additional state.
impl<T> AsyncDrop for Vec<T> {
    async fn drop(&mut self); // ← returns a future
}
```

I don't know whether this redesign would be required just to land *something* in the shorter term. But if we're talking final design, I don't believe we can get away with just `poll_next`.

Reply:
Described there:
https://zetanumbers.github.io/book/async-drop-design.html#choosing-against-poll_drop